File: Python/Matriz.py
class nodo:
    def __init__(self, valor, x =None, y = None):
        self.x = x
        self.y = y
        self.valor = valor
        self.derecho = self.izquierdo = self.arriba = self.abajo = None
        self.siguiente = self.anterior = None

class lista:
    def __init__(self):
        self.raiz = self.ultimo = None

    def insertar(self, valor):
        nuevo = nodo(valor)
        if self.raiz == None:
            self.raiz = self.ultimo = nuevo
        else:
            self.ordenar(nuevo)
    
    def ordenar(self, nodo):
        aux = self.raiz

        while(aux!=None):
            if aux.valor < nodo.valor:
                aux = aux.siguiente
            else:
                if aux == self.raiz:
                    nodo.siguiente = aux
                    aux.anterior = nodo
                    self.raiz = nodo
                    return
                else:
                    nodo.anterior = aux.anterior
                    aux.anterior.siguiente = nodo
                    nodo.siguiente = aux
                    aux.anterior = nodo
                    return
            self.ultimo.siguiente = nodo
            nodo.anterior = self.ultimo
            self.ultimo = nodo
    
    def search(self, valor):
        temp = self.raiz
        while(temp != None):
            if temp.valor == valor:
                return temp
            temp = temp.siguiente
        return None

    def print(self):
        temp = self.raiz
        while(temp!=None):
            print("Cabecera:", temp.valor)
            temp = temp.siguiente
    
class matriz:
    def __init__(self):
        self.lista_horizontal = lista()
        self.lista_vertical = lista()
    
    def insertar(self, valor, x, y):
        horizontal = self.lista_horizontal.search(x)
        vertical = self.lista_vertical.search(y)

        if horizontal == None and vertical == None:
            self.Caso1(valor, x, y)
        elif horizontal == None and vertical != None:
            self.Caso2(valor, x, y)
        elif horizontal != None and vertical == None:
            self.Caso3(valor, x, y)
        else:
            self.Caso4(valor, x, y)

    def Caso1(self, valor, x, y):
        self.lista_horizontal.insertar(x)
        self.lista_vertical.insertar(y)

        horizontal = self.lista_horizontal.search(x)
        vertical = self.lista_vertical.search(y)
        
        nuevo = nodo(valor, x, y)
        horizontal.abajo = nuevo
        nuevo.arriba = horizontal

        vertical.derecho = nuevo
        nuevo.izquierdo = vertical

    def Caso2(self, valor, x, y):
        self.lista_horizontal.insertar(x)

        horizontal = self.lista_horizontal.search(x)
        vertical = self.lista_vertical.search(y)

        agregado = False

        nuevo = nodo(valor, x, y)
        aux = vertical.derecho
        cabecera = 0

        while(aux!= None):
            cabecera = aux.x
            if cabecera < x:
                aux = aux.derecho
            else:
                nuevo.derecho = aux
                nuevo.izquierdo = aux.izquierdo
                aux.izquierdo.derecho = nuevo
                aux.izquierdo = nuevo
                agregado = True
                break
        if agregado == False:
            aux = vertical.derecho
            while(aux.derecho!=None):
                aux = aux.derecho
            nuevo.izquierdo = aux
            aux.derecho = nuevo
        
        horizontal.abajo = nuevo
        nuevo.arriba = horizontal

    def Caso3(self, valor, x, y):
        self.lista_vertical.insertar(y)

        horizontal = self.lista_horizontal.search(x)
        vertical = self.lista_vertical.search(y)

        agregado = False

        nuevo = nodo(valor, x, y)
        aux = horizontal.abajo
        cabecera = 0

        while(aux!= None):
            cabecera = aux.y
            if cabecera < y:
                aux = aux.abajo
            else:
                nuevo.abajo = aux
                nuevo.arriba = aux.arriba
                aux.arriba.abajo = nuevo
                aux.arriba = nuevo
                agregado = True
                break
        if agregado == False:
            aux = horizontal.abajo
            while(aux.abajo!=None):
                aux = aux.abajo
            nuevo.arriba = aux
            aux.abajo = nuevo
        
        vertical.derecho = nuevo
        nuevo.izquierdo = vertical

    def Caso4(self, valor, x, y):
        horizontal = self.lista_horizontal.search(x)
        vertical = self.lista_vertical.search(y)

        nuevo = nodo(valor, x, y)
        
        agregado = False
        aux = horizontal.abajo

        while(aux!= None):
            cabecera = aux.y
            if cabecera < y:
                aux = aux.abajo
            else:
                nuevo.abajo = aux
                nuevo.arriba = aux.arriba
                aux.arriba.abajo = nuevo
                aux.arriba = nuevo
                agregado = True
                break
        if agregado == False:
            aux = horizontal.abajo
            while(aux.abajo!=None):
                aux = aux.abajo
            nuevo.arriba = aux
            aux.abajo = nuevo
        agregado = False
        aux = vertical.derecho
        cabecera = 0

        while(aux!= None):
            cabecera = aux.x
            if cabecera < x:
                aux = aux.derecho
            else:
                nuevo.derecho = aux
                nuevo.izquierdo = aux.izquierdo
                aux.izquierdo.derecho = nuevo
                aux.izquierdo = nuevo
                agregado = True
                break
        if agregado == False:
            aux = vertical.derecho
            while(aux.derecho!=None):
                aux = aux.derecho
            nuevo.izquierdo = aux
            aux.derecho = nuevo
        

    def imprimir_horizontal(self):
        cabecera = self.lista_horizontal.raiz
        while(cabecera != None):
            aux = cabecera.abajo
            while(aux != None):
                print(aux.valor, aux.x, aux.y)
                aux= aux.abajo
            cabecera = cabecera.siguiente

    def imprimir_vertical(self):
        cabecera = self.lista_vertical.raiz
        while(cabecera!=None):
            aux = cabecera.derecho
            while(aux!= None):
                print(aux.valor, aux.x, aux.y)
                aux = aux.derecho
            cabecera = cabecera.siguiente


matriz = matriz()



matriz.insertar(1, 0,0)
matriz.insertar(2, 0,1)
matriz.insertar(3, 1,0)
matriz.insertar(4, 1,1)
print("Cabeceras")
matriz.lista_horizontal.print()
matriz.lista_vertical.print()

print("Se imprime de forma horizontal")
matriz.imprimir_horizontal()
print("Se imprime de forma vertical")
matriz.imprimir_vertical()

File: JavaScript/structure/matriz.js
class Nodo {
    constructor(valor, x, y) {
        this.valor = valor;//Valor con el que vamos a comparar
        this.x = x;//posicion en la cabecera en horizontal
        this.y = y;//posicion en la cabecera en vertical
        this.siguiente = null;//enlace para la cabecera en x
        this.anterior = null;//enlace para la cabera en y
        this.arriba = null;//enlace para moverse hacia arriba en el nodo matriz
        this.abajo = null;//enlace para moverse hacia abajo en el nodo matriz
        this.izquierda = null;//enlace para moverse hacia izquierda en el nodo matriz
        this.derecha = null;//enlace para moverse hacia derecha en el nodo matriz
    }
}

class Lista {
    constructor() {
        this.primero = null;
        this.ultimo = null;
    }

    ordenar(nodo) {
        let aux = this.primero;
        while (aux != null) {
            if (aux.valor < nodo.valor) {
                aux = aux.siguiente;
            } else {
                if (aux == this.primero) {
                    nodo.siguiente = aux;
                    aux.anterior = nodo;
                    this.primero = nodo;
                    return;
                } else {
                    nodo.anterior = aux.anterior;
                    aux.anterior.siguiente = nodo;
                    nodo.siguiente = aux;
                    aux.anterior = nodo;
                    return;
                }
            }
        }
        this.ultimo.siguiente = nodo;
        nodo.anterior = this.ultimo;
        this.ultimo = nodo;
    }

    insertar(valor) {
        let nodo = new Nodo(valor, null, null)
        if (this.primero == null) {
            this.primero = this.ultimo = nodo;
            return;
        }
        this.ordenar(nodo);
    }

    busqueda(valor) {
        let temp = this.primero;
        while (temp != null) {
            if (temp.valor == valor) return temp;
            temp = temp.siguiente;
        }
        return null;
    }
}

class Matriz {
    constructor() {
        this.lista_horizontal = new Lista();
        this.lista_vertical = new Lista();
    }

    insertar(valor, x, y) {
        let nodo_x = this.lista_horizontal.busqueda(x);
        let nodo_y = this.lista_vertical.busqueda(y);

        if (nodo_x == null && nodo_y == null) {
            this.caso1(valor, x, y);
        } else if (nodo_x == null && nodo_y != null) {
            this.caso2(valor, x, y);
        } else if (nodo_x != null && nodo_y == null) {
            this.caso3(valor, x, y);
        } else {
            this.caso4(valor, x, y);
        }
    }

    caso1(valor, x, y) {
        this.lista_horizontal.insertar(x);
        this.lista_vertical.insertar(y);

        let nodo_x = this.lista_horizontal.busqueda(x);
        let nodo_y = this.lista_vertical.busqueda(y);

        let nuevo = new Nodo(valor, x, y);
        nodo_x.abajo = nuevo;
        nuevo.arriba = nodo_x;

        nodo_y.derecha = nuevo;
        nuevo.izquierda = nodo_y;
    }

    caso2(valor, x, y) {
        this.lista_horizontal.insertar(x);

        let nodo_x = this.lista_horizontal.busqueda(x);
        let nodo_y = this.lista_vertical.busqueda(y);

        let agregado = false;

        let nuevo = new Nodo(valor, x, y);
        let aux = nodo_y.derecha;// nos movemos hacia la derecha una posicion
        let cabecera = 0;

        while (aux != null) {
            cabecera = aux.x;
            if (cabecera < x) { // 3 < 4
                aux = aux.derecha;
            } else {
                nuevo.derecha = aux;//aux toma valor de nodo(2)
                nuevo.izquierda = aux.izquierda;
                aux.izquierda.derecha = nuevo;
                aux.izquierda = nuevo;
                agregado = true;
                break;
            }
        }

        if (agregado == false) {
            aux = nodo_y.derecha;
            while (aux.derecha != null) {
                aux = aux.derecha;
            }
            nuevo.izquierda = aux;
            aux.derecha = nuevo;
        }

        nodo_x.abajo = nuevo;
        nuevo.arriba = nodo_x;
    }

    caso3(valor, x, y) {
        // solo existe la cabecera en x
        this.lista_vertical.insertar(y) // inserta la cabecera en y

        let nodo_x = this.lista_horizontal.busqueda(x);
        let nodo_y = this.lista_vertical.busqueda(y);

        let agregado = false;

        let nuevo = new Nodo(valor, x, y);//interno de matriz
        let aux = nodo_x.abajo; // obtenemos el primer nodo de la cabecera
        let cabecera = 0;

        while (aux != null && !agregado) {
            cabecera = aux.y;//1
            if (cabecera < y) {//1<2;2<2;3<2
                aux = aux.abajo;//aux-> nodo(valor:3, x:2, y:3)
            } else {
                nuevo.abajo = aux;
                nuevo.arriba = aux.arriba;
                aux.arriba.abajo = nuevo;
                aux.arriba = nuevo;
                agregado = true;
            }
        }

        if (!agregado) {
            aux = nodo_x.abajo;
            while (aux.abajo != null) {
                aux = aux.abajo;
            }
            aux.abajo = nuevo;
            nuevo.arriba = aux;
        }

        nodo_y.derecha = nuevo;
        nuevo.izquierda = nodo_y;
    }


    caso4(valor, x, y) {

        let nodo_x = this.lista_horizontal.busqueda(x);
        let nodo_y = this.lista_vertical.busqueda(y);

        let agregado = false;
        let nuevo = new Nodo(valor, x, y);
        let aux = nodo_y.derecha;
        let cabecera = 0;
        while (aux != null) {
            cabecera = aux.x;
            if (cabecera < x) {
                aux = aux.derecha;
            } else {
                nuevo.derecha = aux;
                nuevo.izquierda = aux.izquierda;
                aux.izquierda.derecha = nuevo;
                aux.izquierda = nuevo;
                agregado = true;
                break;
            }
        }
        if (agregado == false) {
            aux = nodo_y.derecha;
            while (aux.derecha != null) {
                aux = aux.derecha;
            }
            nuevo.izquierda = aux;
            aux.derecha = nuevo;
        }

        agregado = false;
        aux = nodo_x.abajo;
        cabecera = 0;

        while (aux != null && !agregado) {
            cabecera = aux.y;
            if (cabecera < y) {
                aux = aux.abajo;
            } else {
                nuevo.abajo = aux;
                nuevo.arriba = aux.arriba;
                aux.arriba.abajo = nuevo;
                aux.arriba = nuevo;
                agregado = true;
            }
        }

        if (!agregado) {
            aux = nodo_x.abajo;
            while (aux.abajo != null) {
                aux = aux.abajo;
            }
            aux.abajo = nuevo;
            nuevo.arriba = aux;
        }
    }

    imprimir_vertical(){
        let cabecera = this.lista_vertical.primero;
        let aux;
        while(cabecera != null){
            aux = cabecera.derecha;
            while(aux!=null){
                console.log("Valor:",aux.valor, "X:", aux.x, "Y:", aux.y);
                aux = aux.derecha//iteraciones dentro de la matriz;
            }
            cabecera = cabecera.siguiente;//iteraciones de lista ordenada
        }
    }

    imprimir_horizontal(){
        let cabecera = this.lista_horizontal.primero;
        let aux;
        while(cabecera != null){
            aux = cabecera.abajo;
            while(aux!= null){
                console.log("Valor:",aux.valor, "X:", aux.x, "Y:", aux.y);
                aux = aux.abajo;
            }
            cabecera = cabecera.siguiente;
        }
    }
}

module.exports = Matriz;
